src/Core/Vipr.Core/CodeModel/Vocabularies/Capabilities/OdcmCapability.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Vipr.Core.CodeModel.Vocabularies.Capabilities
+{
+    public abstract class OdcmCapability : IEquatable<OdcmCapability>
+    {
+        public abstract string TermName { get; }
+
+        public abstract bool Equals(OdcmCapability otherCapability);
+
+        public abstract void ParseCapabilityAnnotationForEntitySet(OdcmProperty odcmEntitySet,
+            OdcmVocabularyAnnotation annotation, PropertyCapabilitiesCache propertyCache);
+
+        private static List<OdcmCapability> _defaultOdcmCapabilities;
+
+        /// <summary>
+        /// Default list of OdcmCapabilities supported in the OdcmModel.
+        /// </summary>
+        public static List<OdcmCapability> DefaultOdcmCapabilities
+        {
+            get
+            {
+                if (_defaultOdcmCapabilities == null)
+                {
+                    _defaultOdcmCapabilities = new List<OdcmCapability>();
+                    var capabilityTypes = Assembly.GetExecutingAssembly().GetTypes().Where(t => t.IsSubclassOf(typeof(OdcmCapability)) && !t.IsAbstract);
+
+                    foreach (var capabilityType in capabilityTypes)
+                    {
+                        var capability = (OdcmCapability)Activator.CreateInstance(capabilityType);
+                        _defaultOdcmCapabilities.Add(capability);
+                    }
+                }
+
+                return _defaultOdcmCapabilities;
+            }
+        }
+
+        public static void SetCapabilitiesForEntityContainer(OdcmServiceClass odcmServiceClass)
+        {
+            //TODO: Add Capability Annotation support for EntityContainers
+        }
+
+        public static void SetCapabilitiesForEntitySet(OdcmProperty odcmEntitySet)
+        {
+            if (odcmEntitySet == null) throw new ArgumentNullException("odcmEntitySet");
+
+            var annotations = odcmEntitySet.Annotations;
+            PropertyCapabilitiesCache propertyCache = new PropertyCapabilitiesCache();
+
+            foreach (var annotation in annotations)
+            {
+                var capability = DefaultOdcmCapabilities.SingleOrDefault(c => c.TermName == annotation.FullName);
+                if (capability != null)
+                {
+                    capability.ParseCapabilityAnnotationForEntitySet(odcmEntitySet, annotation, propertyCache);
```

Comment:
In the first commit of the PR I had written the parsing logic in the OdataReader. But that seemed to make the code complicated and a bit difficult to understand since the OdcmCapability and its parser lie in different places. That is why in the second commit I coupled both in one place inside the OdcmCapability class. It made the code look much cleaner and easier to understand.

Also what about "OdcmVocabularyAnnotation" class in OdcmModel? It is used to capture the list of annotations on an OdcmObject. Isn't this specific to OData?